### ISSUE TITLE
ENH: Flatten PDF forms

### ISSUE THREAD

**Issue by OpenNingia** [py-pdf/pypdf#232]
pdftk provides the feature to embed the form fields' text in the pdf itself. 
This is very useful if you want to use an editable pdf as a template to be filled by code.

from the pdftk manual:

```
[ flatten ]
Use this option to merge an input PDF’s interactive form fields (and their data) with the PDF’s pages. Only one input PDF can be given. Sometimes used with the fill_form operation.
```

usage example:

``` python
    with open(source, 'rb') as source_fp:
        reader = PdfFileReader(source_fp)

        writer.appendPagesFromReader(
            reader, lambda x: writer.updatePageFormFieldValues(x, fields))

        with open(dest, 'wb') as output_fp:
            writer.write(output_fp, flatten_fields=True)

```

**Comment by whitemice** [comment]
+1  A way to flatten a form would be excellent.  I would like to avoid having another dependency for my code, which uses PyPDF2.  But shipping filled in forms around the interwebz creates problems with a variety of vendors and their [I assume not based on PyPDF2] software.

**Comment by mertz3hack** [comment]
It would be great if PyPDF2 had the ability to fill in forms and flatten them!

**Comment by oscardssmith** [comment]
I also would really appreciate this

**Comment by mstamy2** [comment]
(In progress) We can accomplish this by setting Bit Position 1 of the field flags.

**Comment by OpenNingia** [comment]
Setting a field read-only might be a way, however pdftk works differently; afaik it replaces each /Fields instance with a simple text object. :confused:

**Comment by mstamy2** [comment]
You're right, that's the better option. Should be able to implement that soon

**Comment by nberrios** [comment]
I agree. This would be totally awesome!

**Comment by jamoham** [comment]
Is there any update on this?
I am looking to use an editable pdf as a template which will be filled by code.

**Comment by kherrett** [comment]
I'm with @jamoham on this... for the same exact use case.

**Comment by zhiwehu** [comment]
+1

**Comment by Rob1080** [comment]
Any update on this?

**Comment by BeGrimm** [comment]
Can you flatten a file with PyPDF2 yet?  I've not found anything on this being implemented.

**Comment by DrLou** [comment]
I do see some code to __flatten_ in the PdfFileReader, but not in the writer. Will someone be taking a swing at this?

**Comment by Joshua-IRT** [comment]
I have exactly the same scenario as mentioned by @jamoham, @kherrett and @zhiwehu above. Has there been any progress on either being able to flatten a PDF, or set the fields as read-only?

**Comment by Joshua-IRT** [comment]
Rough bit of code if anyone needs to set fields to read-only prior to an update to the module (assumes you imported the whole module as PyPDF2). Works in a similar fashion to the existing `updatePageFormFieldValues()` method.

```
class PDFModifier(PyPDF2.PdfFileWriter):
    '''Extends the PyPDF2.PdfFileWriter class and adds functionality missing
    from the PyPDF2 module.'''

    def updatePageFormFieldFlags(self, page, fields, or_existing=True):
        '''
        Update the form field values for a given page from a fields dictionary.
        Copy field texts and values from fields to page.

        :param page: Page reference from PDF writer where the annotations
            and field data will be updated.
        :param fields: a Python dictionary of field names (/T) and flag
            values (/Ff); the flag value should be an unsigned 32-bit integer
            (i.e. a number between 0 and 4294967295)
        :param or_existing: if there are existing flags, OR them with the
            new values (default True)
        '''

        # Iterate through pages and update field flag
        for j in range(0, len(page['/Annots'])):
            writer_annot = page['/Annots'][j].getObject()
            for field in fields:
                if writer_annot.get('/T') == field:
                    if or_existing:
                        current_flags = writer_annot.get('/Ff')
                        if current_flags is not None:
                            fields[field] = int(bin(current_flags | fields[field]),2)

                    writer_annot.update({
                        PyPDF2.generic.NameObject("/Ff"): PyPDF2.generic.NumberObject(fields[field])
                    })
```

**Comment by chickendiver** [comment]
+1 for flattening, such as in pdftk!

**Comment by techNoSavvy-debug** [comment]
+1 for a method for flattening pdfs

**Comment by paulzuradzki** [comment]
One thing I noticed with the approach of flattening/making forms read-only by setting the field flag bit to 1: when I try to merge resulting PDFs, only the values from the first document make it to the merged file.

pdftk does not seem to have this issue with its approach to flattening.

**Comment by paulzuradzki** [comment]
Cross-posting this useful recipe by @Redjumpman. Remember to update the form field name if you want to merge multiple documents made from the same template form. Else, the merged PDF result will have identical pages due to each document sharing the same field names.

https://github.com/mstamy2/PyPDF2/issues/506

**Comment by pubpub-zz** [comment]
`PdfWriter.append()` should provide you with capability to add pages with data fields.

Can you confirm that this issue can get closed?

**Comment by pubpub-zz** [comment]
without feed back I close this issue as fixed. Feel free to provides updates if yuo wan to reopen it.

**Comment by pubpub-zz** [comment]
the read-only flag defined here in the Pdf 1.7 reference (page 676)
![image](https://user-images.githubusercontent.com/4083478/225142764-74a8842e-b635-4209-b084-dc82ea341136.png)

therefore you have to set the flags. Below an example setting all the fields in readonly:
```
import pypdf
r = pypdf.PdfReader("input_form.pdf")
for f,v in r.get_fields().items():
  o=v.indirect_reference.get_object()   # this will provide access to the actual PDF dictionary 
  o[NameObject("/Ff")] = NumberObject( o.get("/Ff",0)|1)
w = pypdf.PdfWriter()
w.clone_document_from_reader(r)
w.write("output_form.pdf")
```

**Comment by OpenNingia** [comment]
What you are suggesting is not "flattening" thou. The output pdf will still present data fields.
Flattening as pdftk does is replacing the data field with text.

**Comment by pubpub-zz** [comment]
@OpenNingia 
Can you provide a pdf non flatten and its flatten version for review

**Comment by pubpub-zz** [comment]
The flattening process  is quite tough to compute (create XOBject with the good characteristics) modify the content to place them.
I see personnally very limited advantage vs time to implement an for me the readonly alternative could be sufficient ; I will have no time to propose a PR. Any candidate ?

**Comment by pubpub-zz** [comment]
since we have now #1864, flattening should be quite simple

**Comment by matsavage** [comment]
I have subclassed the `PdfWriter` class to be able to flatten forms here, so it can be done, would you accept PR for this, and do you have any idea of the interface which would be best for implementation, I think this would be the easiest option, or there could be something more advanced, where a list of annotations or None is passed, then the specified annotations, or all are flattened, but would be hesitant to do something too elaborate to avoid running before walking.

https://gist.github.com/matsavage/a50d9c541957f276088c341cc84a9e7f

**Comment by matsavage** [comment]
I only did things this way to see if the flattening could be done, to save the effort of setting up the development environment on my machine, this is more the template than the PR

Thanks for the advice, I’ll try and have a look at this some time

**Comment by pubpub-zz** [comment]
Looking forward 😊

**Comment by nicholas-alonzo** [comment]
Looking forward for this feature!

**Comment by matsavage** [comment]
Honestly I haven’t been able to look at this since May, feel free to have your own attempt at implementing it if it’s something you need.

**Comment by pubpub-zz** [comment]
At your marks.... get set ... go! 😉😄😄😄

**Comment by matsavage** [comment]
> At your marks.... get set ... go! 😉😄😄😄

I think it’s the one everyone wants, but no one wants to do

**Comment by nicholas-alonzo** [comment]
> Honestly I haven’t been able to look at this since May, feel free to have your own attempt at implementing it if it’s something you need. 

Darn, I wouldn't even know where to start 🥴

**Comment by PJBrs** [comment]
I've used google gemini to come up with flattening code for text fields and some buttons. I've very basically followed @matsavage 's lead by subclassing PdfWriter and then adding a flatten function along with some helpers. For text fields and buttons, it creates appearance streams and adds these as XObject Forms. I've made this part of an existing project, but I think anyone should be able to just include the following file and include the CustomPdfWriter from it:

https://github.com/PJBrs/dungeon-sheets/blob/flatten_vibe/dungeonsheets/flatten_pdf.py

It does include a dependency on pdfminer.six to calculate font widths (pdfminer.six has a table with character widths for the 14 base pdf fonts).

I'm sure this can be improved by someone who knows something about pdfs. Also, I think there's quite a bit redundant safety checks. Otherwise, I'm curious if this code works for others as well!

**Comment by stefan6419846** [comment]
We are still open to include this into *pypdf* itself to avoid having to maintain external solutions. PRs are highly appreciated.

**Comment by PJBrs** [comment]
@stefan6419846 I opened a PR (https://github.com/py-pdf/pypdf/pull/3312). I think quite some review is needed, and maybe, to begin with, a couple of concrete leads for how to proceed.
